User Says addition (#548)
hellohaptik/docs_haptik: docs/bot-builder/basic/user-says.md

title: User Says
---

> # Introduction

Reviewer: There is a similar introduction on User Says Guidelines doc, please check that and remove the repeating bits @rasika-purohit

Whenever a user sends a message, the bot tries to understand what the User is trying to say, using various Machine Learning algorithms and find the corresponding node.

> 

Reviewer: Editing and deactivating actions on User Says should also be explained

Show services | Tell me about demat account

> **Tip**: User Responses are case insensitive and ignore punctuations.
> 
>  So *don’t* and *dONt* are the same.

## Negative Responses

The above section explained how you can select a node depending on what the user said. This section will explain how to filter out nodes depending on what the user says.

For example, if you have the following nodes:

**Good Service** | **Bad Service**
-----------------| ---------------
Your service was *very* good | Your service was **not** good.
Your service was excellent | Your service was bad

Now when a user says "Your service was not good", even though the User meant something very different, because of high similarity of the sentences in both the nodes, it is possible that an Incorrect node may get selected.

To avoid this, you can add specific keywords under **Negative Responses** of the node which will filter out the node, if the keyword is a part of the user's message.

For Example, in the above scenario, if you add "bad" as a Negative Response on the "Good Service" node, then the Good Service node will not be considered if the User's message has "bad" in it's sentence.

> > Adding similar responses on 2 different nodes

Reviewer: I did not understand this node @rasika-purohit

> 

Reviewer: We need to add something here which would describe what are we going to talk next. For now it looks quite random. @rasika-purohit

## Train Bot
The User Responses added on the node above are processed using various Machine Learning algorithms to generate the models. These models are then used on User’s every message to select the node which matches the requirements.

![image](https://user-images.githubusercontent.com/75118325/124480900-5dfba200-ddc5-11eb-80d1-23a7a0aeefa8.png)

To generate these models, you will have to click on **Train** Bot. This will send an email with the result of the training process. It will also contain details of any probable errors which could come up because of incorrectly entered data.

For example, if you add very similar sentences in two different nodes, then the resulting email will show details about the conflicting nodes and responses.

> **Tips**: Check [**Oracle feedback**](https://docs.haptik.ai/bot-builder/basic/oracle-feedback) email for conflicting responses or other errors in the User Responses in the bot.

## Autocomplete suggestions 

For all bots on Web SDK, as users start typing, the Taskbox can turn dynamic in autocomplete suggestions mode (refer the image below). The Machine Learning model matches the user's typed letters with user says sentences on the bot. 

>       You have to enable autcomplete feature manually on your bot. 

Reviewer: We should add where to enable this option @rasika-purohit

![Marking Autocomplete suggestions](assets/bot-builder-user-says/user-says.png)

**Step 2** - In the Business Manager enable to **Enable Typing Suggestions** option from the **SDK Configurations** section.

![image](https://user-images.githubusercontent.com/75118325/124481372-d4000900-ddc5-11eb-88f3-547aa026ff76.png)

This feature helps to -

> 1. Help users discover the scope of our bot while an user starts typing

Reviewer: ```suggestion
1. Discover the usecases on the bot while a user starts typing
```

> 2. Avoid bot breaks as user selects our suggestion than type a random freeform message.

Reviewer: ```suggestion
2. Reduce bot breaks as user selects the suggestion than type a random freeform message.
```
